src/odemis/gui/comp/overlay/shapes.py

```diff
@@ -237,9 +288,73 @@ def on_left_up(self, evt):
         self._selected_shape = self._get_shape(evt)
         if self._selected_shape:
             self._selected_shape.on_left_up(evt)
+            state = self._selected_shape.get_state()
+            if state:
+                shape_state = ShapeState(self._selected_shape, state, self._is_new_shape, False)
+                if not self._undo_stack or self._undo_stack[-1] != shape_state:
+                    self._undo_stack.append(shape_state)
+                    self._redo_stack.clear()  # Clear redo stack when a shape's state is saved
         else:
             WorldOverlay.on_left_up(self, evt)
 
+    def undo(self):
+        """Undo the last action."""
+        if self._undo_stack:
+            # If the redo stack is empty or if the previous action was redo
+            # correct the off by 1 shape's state by popping it and not restoring the state
+            if not self._redo_stack or self._redo_action:
+                self._redo_action = False
+                shape_state = self._undo_stack.pop()
+                self._redo_stack.append(shape_state)
+                if shape_state.is_new:
+                    self.remove_shape(shape_state.shape)
+                    return
+                elif shape_state.is_deleted:
+                    self.add_shape(shape_state.shape)
+                    return
+            if self._undo_stack:
+                shape_state = self._undo_stack.pop()
+                self._redo_stack.append(shape_state)
+                shape_state.shape.restore_state(shape_state.state)
+                if shape_state.is_new:
+                    self.remove_shape(shape_state.shape)
+                elif shape_state.is_deleted:
+                    self.add_shape(shape_state.shape)
+                self.cnvs.request_drawing_update()
+        else:
+            logging.info(
+                "No undo action for %s at %s.", self.__class__.__name__, hex(id(self))
+            )
+
+    def redo(self):
+        """Redo the last undone action."""
+        if self._redo_stack:
+            # If the undo stack is empty or if the previous action was undo
+            # correct the off by 1 shape's state by popping it and not restoring the state
+            if not self._undo_stack or self._undo_action:
```

Comment:
I don't think the `not undo_stack` is useful here. That's because undo/redo are not symmetrical: editing a shape reset just the redo_stack. So the only way to have an empty undo_stage is that the last action was an undo_action. Hence, it's not useful here.